store/sqlstore/compliance_store.go

```diff
@@ -209,8 +210,15 @@ func (s SqlComplianceStore) ComplianceExport(job *model.Compliance) ([]*model.Co
 
 func (s SqlComplianceStore) MessageExport(after int64, limit int) ([]*model.MessageExport, error) {
 	props := map[string]interface{}{"StartTime": after, "Limit": limit}
-	query :=
-		`SELECT
+
+	var explictUpdateAtIndex string
+
+	if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		explictUpdateAtIndex = "FORCE INDEX (idx_posts_update_at)"
```

Comment:
> I am wondering whether we still need this given the fact that now we have simplified the query. Is this out of abundance of caution or did you experience MySQL doing something wonky in your test setup?
> 
> I'd usually prefer to leave index decisions to the DB, until we have evidence that it is deliberately making the wrong choice. For example, if the posts table is quite small, then it might be faster to do a table-scan than an index based scan.

This is what I'm getting on my system, and why I had kept it:

# Cold cache

## With `FORCE INDEX (idx_posts_update_at)`: 1:51

```
-> Limit: 200000 row(s)  (actual time=13.642..110261.290 rows=200000 loops=1)
    -> Nested loop left join  (cost=16694527.60 rows=4456941) (actual time=13.641..110220.600 rows=200000 loops=1)
        -> Nested loop left join  (cost=11791892.55 rows=4456941) (actual time=10.596..109997.601 rows=200000 loops=1)
            -> Nested loop left join  (cost=11334658.18 rows=4456941) (actual time=10.583..108695.995 rows=200000 loops=1)
                -> Nested loop left join  (cost=10880076.10 rows=4456941) (actual time=10.569..107251.881 rows=200000 loops=1)
                    -> Filter: (not((posts.`Type` like 'system_%')))  (cost=5977441.06 rows=4456941) (actual time=10.507..105554.534 rows=200000 loops=1)
                        -> Index range scan on Posts using idx_posts_update_at, with index condition: (posts.UpdateAt > 1600280586438)  (cost=5977441.06 rows=5013996) (actual time=0.826..105080.095 rows=200011 loops=1)
                    -> Single-row index lookup on Channels using PRIMARY (Id=posts.ChannelId)  (cost=1.00 rows=1) (actual time=0.008..0.008 rows=1 loops=200000)
                -> Filter: (teams.Id = channels.TeamId)  (cost=0.00 rows=1) (actual time=0.006..0.007 rows=1 loops=200000)
                    -> Table scan on Teams  (cost=0.00 rows=1) (actual time=0.005..0.005 rows=1 loops=200000)
            -> Filter: (users.Id = posts.UserId)  (cost=0.00 rows=1) (actual time=0.005..0.006 rows=1 loops=200000)
                -> Table scan on Users  (cost=0.00 rows=1) (actual time=0.004..0.005 rows=1 loops=200000)
        -> Single-row index lookup on Bots using PRIMARY (UserId=posts.UserId)  (cost=1.00 rows=1) (actual time=0.001..0.001 rows=0 loops=200000)
```

## No explicit index: 8:29

```
-> Limit: 200000 row(s)  (actual time=508505.945..508648.325 rows=200000 loops=1)
    -> Sort: posts.UpdateAt, limit input to 200000 row(s) per chunk  (actual time=508505.905..508634.642 rows=200000 loops=1)
        -> Stream results  (actual time=12.681..438966.052 rows=10904763 loops=1)
            -> Nested loop left join  (cost=5359869.27 rows=4456941) (actual time=12.663..414560.150 rows=10904763 loops=1)
                -> Left hash join (users.Id = posts.UserId)  (cost=457234.37 rows=4456941) (actual time=9.499..408621.694 rows=10904763 loops=1)
                    -> Left hash join (teams.Id = channels.TeamId)  (cost=454582.08 rows=4456941) (actual time=9.440..402628.073 rows=10904763 loops=1)
                        -> Nested loop left join  (cost=6034939.82 rows=4456941) (actual time=9.406..392318.561 rows=10904763 loops=1)
                            -> Filter: (not((posts.`Type` like 'system_%')))  (cost=1132304.93 rows=4456941) (actual time=9.354..359288.265 rows=10904763 loops=1)
                                -> Index range scan on Posts using idx_posts_update_at, with index condition: (posts.UpdateAt > 1600280586438)  (cost=1132304.93 rows=5013996) (actual time=0.890..353798.915 rows=10904774 loops=1)
                            -> Single-row index lookup on Channels using PRIMARY (Id=posts.ChannelId)  (cost=1.00 rows=1) (actual time=0.003..0.003 rows=1 loops=10904763)
                        -> Hash
                            -> Table scan on Teams  (cost=0.00 rows=1) (actual time=0.008..0.009 rows=1 loops=1)
                    -> Hash
                        -> Table scan on Users  (cost=0.00 rows=1) (actual time=0.024..0.026 rows=1 loops=1)
                -> Single-row index lookup on Bots using PRIMARY (UserId=posts.UserId)  (cost=1.00 rows=1) (actual time=0.000..0.000 rows=0 loops=10904763)
```

# Hot cache

## With `FORCE INDEX (idx_posts_update_at)`: 0:05

```
-> Limit: 200000 row(s)  (actual time=0.291..5119.730 rows=200000 loops=1)
    -> Nested loop left join  (cost=13192205.24 rows=4456941) (actual time=0.290..5101.023 rows=200000 loops=1)
        -> Nested loop left join  (cost=8289570.19 rows=4456941) (actual time=0.241..4990.422 rows=200000 loops=1)
            -> Nested loop left join  (cost=7832335.82 rows=4456941) (actual time=0.235..4401.546 rows=200000 loops=1)
                -> Nested loop left join  (cost=7377753.74 rows=4456941) (actual time=0.228..3821.712 rows=200000 loops=1)
                    -> Filter: (not((posts.`Type` like 'system_%')))  (cost=5817824.41 rows=4456941) (actual time=0.215..3323.212 rows=200000 loops=1)
                        -> Index range scan on Posts using idx_posts_update_at, with index condition: (posts.UpdateAt > 1600280586438)  (cost=5817824.41 rows=5013996) (actual time=0.036..3249.674 rows=200011 loops=1)
                    -> Single-row index lookup on Channels using PRIMARY (Id=posts.ChannelId)  (cost=0.25 rows=1) (actual time=0.002..0.002 rows=1 loops=200000)
                -> Filter: (teams.Id = channels.TeamId)  (cost=0.00 rows=1) (actual time=0.002..0.003 rows=1 loops=200000)
                    -> Table scan on Teams  (cost=0.00 rows=1) (actual time=0.002..0.002 rows=1 loops=200000)
            -> Filter: (users.Id = posts.UserId)  (cost=0.00 rows=1) (actual time=0.002..0.003 rows=1 loops=200000)
                -> Table scan on Users  (cost=0.00 rows=1) (actual time=0.002..0.002 rows=1 loops=200000)
        -> Single-row index lookup on Bots using PRIMARY (UserId=posts.UserId)  (cost=1.00 rows=1) (actual time=0.000..0.000 rows=0 loops=200000)
```

## No explicit index: 5:22

```
-> Limit: 200000 row(s)  (actual time=321914.284..322043.657 rows=200000 loops=1)
    -> Sort: posts.UpdateAt, limit input to 200000 row(s) per chunk  (actual time=321913.463..322028.298 rows=200000 loops=1)
        -> Stream results  (actual time=21.112..281907.017 rows=10904763 loops=1)
            -> Nested loop left join  (cost=5394490.11 rows=4456941) (actual time=21.077..258631.257 rows=10904763 loops=1)
                -> Left hash join (users.Id = posts.UserId)  (cost=491855.21 rows=4456941) (actual time=17.944..252995.781 rows=10904763 loops=1)
                    -> Left hash join (teams.Id = channels.TeamId)  (cost=481246.03 rows=4456941) (actual time=13.521..247235.854 rows=10904763 loops=1)
                        -> Nested loop left join  (cost=2688121.34 rows=4456941) (actual time=2.143..237513.122 rows=10904763 loops=1)
                            -> Filter: (not((posts.`Type` like 'system_%')))  (cost=1128192.05 rows=4456941) (actual time=2.059..207338.168 rows=10904763 loops=1)
                                -> Index range scan on Posts using idx_posts_update_at, with index condition: (posts.UpdateAt > 1600280586438)  (cost=1128192.05 rows=5013996) (actual time=0.330..203001.946 rows=10904774 loops=1)
                            -> Single-row index lookup on Channels using PRIMARY (Id=posts.ChannelId)  (cost=0.25 rows=1) (actual time=0.002..0.002 rows=1 loops=10904763)
                        -> Hash
                            -> Table scan on Teams  (cost=0.01 rows=1) (actual time=11.333..11.337 rows=1 loops=1)
                    -> Hash
                        -> Table scan on Users  (cost=0.01 rows=1) (actual time=4.267..4.271 rows=1 loops=1)
                -> Single-row index lookup on Bots using PRIMARY (UserId=posts.UserId)  (cost=1.00 rows=1) (actual time=0.000..0.000 rows=0 loops=10904763)
```

Reply:
Very strange indeed. I see both queries using the index,  but the force index one is somehow accurately giving the rows in a pre-sorted fashion due to which all of the joins with the others tables are nicely done with 200000 rows, and you don't even need to sort at the end. I cannot explain this behavior. Any thoughts?

Reply:
My initial hypothesis is that the build phase of the hash join algorithm—when MySQL chooses it—needs to look at more rows because the input `Posts` table is so big. I'm going to dig in more to properly understand what's happening and under what conditions it would be slower to force the index (for example, if the LIMIT is over a particular size, or if the other tables being joined are bigger, etc).

Reply:
Perfect, thanks for that Martin. Appreciate it.